Support all combinations of datatypes and transposes/adjoints in LinearAlgebra

martenlienen opened the pull request:
This PR generalizes the LinearAlgebra integration of CUSPARSE from a handpicked list of common types to a large list of combinations of linear algebra types.The definitions are automatically generated to avoid "holes" in the methods.

I ran into this today when I was using Flux with Zygote. The pullback of matrix multiplication is defined as `A' * \Delta` which was not defined in CUDA.jl previously.

Do I need to register the new test file somewhere so that it runs in the CI? The `runtests.jl` is pretty elaborate and I could not figure out if it would pick up the file on its own or not.

martenlienen committed:
Support all combinations of datatypes and transposes/adjoints

maleadt commented:
Thanks! Tests are picked up automatically, it scans for files in the `tests` directory. Please rebase on top of master when https://github.com/JuliaGPU/CUDA.jl/pull/532 is merged, that should fix current CI issues.
EDIT: I rebased already.

martenlienen committed:
Check for supported op combinations in CUDA10

martenlienen committed:
Fail loudly instead of silently doing the wrong thing

Previously, mv! and mm! with the adjoint of a complex CSC matrix would just silently take
the transpose, foregoing the complex conjugation, leading to wrong results.

martenlienen commented:
I think the CI errors on 10.1 and 10.2 are a bug that is unrelated to this PR, see the following example.
```julia
ENV["JULIA_CUDA_VERSION"] = "10.2"
using CUDA, CUDA.CUSPARSE, SparseArrays, LinearAlgebra
CUDA.versioninfo()

n = 2
elty = Float32
dA = CuSparseMatrixCSR(elty.(sparse(I, n, n)))
dB = CuArray(reshape(elty.(1:n^2), n, n))
dC = similar(dB)

# Multiplying B with an identity matrix works as expected
CUSPARSE.mm!('N', 'N', 1, dA, dB, 0, dC, 'O')
# 2×2 CuArray{Float32,2}:
#  1.0  3.0
#  2.0  4.0

# Transposing B leads to an indexing problem most likely that makes CUDA read uninitialized memory
CUSPARSE.mm!('N', 'T', 1, dA, dB, 0, dC, 'O')
# 2×2 CuArray{Float32,2}:
#  2.0  3.0
#  4.0  5.7f-44

# The same if we take the adjoint
CUSPARSE.mm!('N', 'C', 1, dA, dB, 0, dC, 'O')
# The 
# 2×2 CuArray{Float32,2}:
#  2.0  3.0
#  4.0  5.7f-44

# Transposing/adjointing A as well makes everything somehow work again?!
CUSPARSE.mm!('T', 'C', 1, dA, dB, 0, dC, 'O')
# 2×2 CuArray{Float32,2}:
#  1.0  2.0
#  3.0  4.0
```
Maybe that is even a bug in CUDA itself? At least I could not find anything in CUDA.jl that would make this happen. Do you have an idea? Should I exclude these test cases on CUDA versions older than 11?

I have also added another commit that makes the program explode instead of silently producing wrong results in a scenario that I ran into when I first wrote the tests with CSC matrices. This way at least hopefully, the next person does not spend half an hour figuring this out.

martenlienen commented:
@maleadt Do you have an opinion on how I should handle this? The problem for me is that I cannot find something that changed between 10.2 and 11 to make these tests fail and CUDA11 seems to agree that this should just work.

maleadt commented:
Sorry, forgot about this. Can we use older, non-generic level 3 APIs for this version of CUDA instead? We already do something like it for 10.1. If not, I'd mark the test as broken and add a warning/error when users try to use the functionality.

martenlienen commented:
And finally everything is green :) I tried out `mm2!` and it works for `B^T` and `A` but no other case. So I went with that solution and added error messages for the other combinations (`A^T, B^T`, `A, B^H` etc.) when you are running on CUDA10.

maleadt commented on test/cusparse.jl:
```diff
@@ -711,17 +707,15 @@ end
-            @test_throws DimensionMismatch mm!('T','T',alpha,d_A,d_B,beta,d_C,'O')
+            if !(CUSPARSE.version() < v"10.3.1" && d_A isa CuSparseMatrixCSR)
+                # A^T is not supported with B^T
+                @test_throws DimensionMismatch mm!('T','T',alpha,d_A,d_B,beta,d_C,'O')
```
I wouldn't throw a DimensionMismatch here, but an `error` that indicates this is related to the CUDA version.

maleadt commented on lib/cusparse/generic.jl:
```diff
@@ -100,8 +100,10 @@ function mv!(transa::SparseChar, alpha::Number, A::CuSparseMatrixCSC{T}, X::Dens
     if transa == 'N'
         ctransa = 'T'
+    elseif transa == 'C' && T <: Complex
+        error("Matrix-vector multiplication with the adjoint of a CSC matrix is not " *
```
Maybe `ArgumentError` for this and the other `error` calls?

martenlienen replied:
Yes, makes sense. I have changed all `error` into `throw(ArgumentError(...))`.

maleadt reviewed (changes requested):
Great, thanks for finishing this! Some final comments, but this should be good to go.

martenlienen commented:
I have fixed the merge conflicts but buildkite seems to have a problem with load, maybe? In the last two runs both times different machines timed out. Let's hope third time's the charm.

martenlienen commented:
@maleadt I have now given this a fourth try and one of the jobs got stuck again. Do you have an idea? These boxes always get to here
```
Downloading artifact: CUDA110
Downloading artifact: CUDNN_CUDA110
   Downloading [=>                                       ]  0.2 %
```
and then they do not progress any further until buildkite interrupts them at some point. This time it was CUDA11.0 but I think it is just random.

maleadt commented:
I couldn't reproduce locally, but disabled use of the package server on `master` so try rebasing again. If it still fails, there's probably something up with the new Downloads.jl-based functionality on Julia 1.6. cc @staticfloat

martenlienen commented:
Now it worked, so this should finally be good to go :)

maleadt commented:
Great, thanks!
